kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiBatchApplicationStream.java

```diff
@@ -88,27 +91,83 @@ public void preShutdown() {
      */
     @Override
     public void appendRows(TableName tableName, List<Object[]> rows, String streamName) {
-        JSONArray jsonRecords = new JSONArray();
-        rows.forEach(item -> jsonRecords.put(item[1]));
+        JSONArray jsonRecords;
+        StorageWriteApiRetryHandler retryHandler = new StorageWriteApiRetryHandler(tableName, getSinkRecords(rows), retry, retryWait);
+        logger.debug("Sending {} records to write Api Application stream {} ...", rows.size(), streamName);
         ApplicationStream applicationStream = this.streams.get(tableName.toString()).get(streamName);
-        try {
-            ApiFuture<AppendRowsResponse> response = applicationStream.writer().append(jsonRecords);
-            applicationStream.increaseAppendCall();
-            AppendRowsResponse writeResult = response.get();
 
-            if (writeResult.hasAppendResult()) {
-                logger.trace("Append call completed successfully on stream {}", applicationStream.getStreamName());
-                applicationStream.increaseCompletedCalls();
-                commitStreamIfEligible(tableName.toString(), applicationStream.getStreamName());
-            } else if (writeResult.hasUpdatedSchema()) {
-                //TODO: Schema Update attempt once
-            } else if (writeResult.hasError()) {
-                //TODO: exception handling
+        do {
+            try {
+                jsonRecords = new JSONArray();
+                for (Object[] item : rows) {
+                    jsonRecords.put(item[1]);
+                }
+                if (retryHandler.getAttempt() == 0) {
+                    // We only consider 1 attempt for 1 batch of requests until it is successful or fails completely
+                    applicationStream.increaseAppendCall();
+                }
+
+                logger.trace("Sending records to Storage API writer for batch load...");
+                ApiFuture<AppendRowsResponse> response = applicationStream.writer().append(jsonRecords);
+                AppendRowsResponse writeResult = response.get();
```

Comment:
we are treating async api as a synchronous call by invoking a get immediately. just a point to note.